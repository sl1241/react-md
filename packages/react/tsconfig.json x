{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "declarationDir": "./dist/types",
    "composite": true,
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "jsx": "preserve",
    "jsxImportSource": "voby",
    "types": [
      "voby/dist/jsx/jsx"
    ],
  },
  "exclude": [
    "vite.config.ts",
"vite.config.umd.ts",
    "types",
  ],
}
