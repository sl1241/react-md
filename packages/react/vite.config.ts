import { defineConfig } from 'vite'

const config = defineConfig({
    build: {
        minify: false,
        lib: {
            entry: ["./src/index.ts"],
            name: "react",
            formats: ['cjs', 'es', 'umd'],
            fileName: (format: string, entryName: string) => `${entryName}.${format}.js`
        },
        // rollupOptions: {
        //   output: {
        //     preserveModules: true,
        //   }
        // },
    },


    esbuild: {
        jsx: 'automatic',
    },

    css: {
        modules: {
            scopeBehaviour: 'global',
            localsConvention: 'camelCase',
        }
    },
})



export default config
