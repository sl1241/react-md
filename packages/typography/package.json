{
  "name": "@react-md/typography",
  "version": "5.1.3",
  "description": "The base package for including typography for react-md.",
  "main": "./lib/index.js",
  "module": "./es/index.js",
  "types": "./types/index.d.ts",
  "sideEffects": [
    "dist/**/*"
  ],
  "author": "Mikkel Laursen <mlaursen03@gmail.com>",
  "repository": {
    "type": "git",
    "url": "https://github.com/mlaursen/react-md.git",
    "directory": "packages/typography"
  },
  "bugs": {
    "url": "https://github.com/mlaursen/react-md/issues"
  },
  "homepage": "https://react-md.dev/packages/typography/demos",
  "keywords": [
    "react-md",
    "material design",
    "react",
    "component",
    "components",
    "typography"
  ],
  "license": "MIT",
  "dependencies": {
    "@react-md/theme": "workspace:../theme",
    "@react-md/utils": "workspace:../utils",
    "classnames": "^2.3.2"
  },
  "devDependencies": {
    "@types/react": "^18.0.29",
    "csstype": "^3.1.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "peerDependencies": {
    "react": ">= 16.14",
    "react-dom": ">= 16.14"
  },
  "peerDependenciesMeta": {
    "react": {
      "optional": false
    },
    "react-dom": {
      "optional": false
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
