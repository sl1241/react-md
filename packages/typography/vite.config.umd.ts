import { defineConfig } from 'vite'

const config = defineConfig({
    build: {
        minify: false,
        lib: {
            entry: ["./src/index.ts"],
            name: "typography",
            formats: ['umd'],
            fileName: (format: string, entryName: string) => `${entryName}.${format}.js`
        },
        emptyOutDir: false,
        // rollupOptions: {
        //   output: {
        //     preserveModules: true,
        //   }
        // },
    },


    esbuild: {
        jsx: 'automatic',
    },
})


export default config
