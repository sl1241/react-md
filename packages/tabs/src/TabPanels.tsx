
import {
    cloneElement,
    $,
    useEffect,
    $$,
}from 'voby'

import { bem, useEnsuredRef } from "@react-md/utils"
import { Children } from '@react-md/react'

import { useTabs } from "./TabsManager"
import type { TabPanelProps } from "./TabPanel"

export interface TabPanelsProps<T extends EventTarget = HTMLDivElement> extends HTMLAttributes<T> {
    /**
     * Boolean if this component should no longer automatically reset the scrolling
     * to the top when the panel changes.
     *
     * @defaultValue `false`
     */
    disableScrollFix?: FunctionMaybe<Nullable<boolean>>

    /**
     * Boolean if the swiping transition should be disabled. If you want to add
     * a custom transition, you'll need to wrap the `TabPanel`'s children in a
     * custom component that does appear and exit animations.
     *
     * @defaultValue `false`
     */
    disableTransition?: FunctionMaybe<Nullable<boolean>>

    /**
     * Boolean if the conditional rendering for the active tab panel only should
     * be disabled. This means that all the children will be visible in the DOM
     * instead of mounting and unmounting when their active state changes. The
     * panels will also be updated to ensure that inactive panels can not be
     * tab focusable.
     *
     * @defaultValue `false`
     */
    persistent?: FunctionMaybe<Nullable<boolean>>
}

const block = bem("rmd-tab-panels")

/**
 * This component allows you to control the visibility of the `TabPanel`
 * components and animating the next and current panels as needed. This works by
 * looping over all the children and getting the current `TabPanel` by the
 * `activeIndex`. This is why the children for this component can only be
 * `TabPanel` and should not be conditional.
 */
export const TabPanels = (
    {
        className,
        children,
        disableScrollFix = false,
        disableTransition = false,
        persistent = false,
ref:        forwardedRef,
        ...props
    }: TabPanelsProps<HTMLDivElement>
) =>{
    const { tabsId, tabs, activeIndex } = useTabs()
    const prevIndex = $(activeIndex)
    const state = $({
        previous: activeIndex,
        incrementing: true,
    })
    const { previous, incrementing }=state()


    // have to set these in refs since changing these might cause mounting
    // and unmounting in the Transition group component :/ they should only
    // be re-evaluated when the activeIndex changes.
    const transitionable = $(!persistent && !disableTransition)
    const animatable = $(persistent && !disableTransition)
    if (prevIndex(activeIndex)) {
        prevIndex(activeIndex)
        transitionable(!persistent && !disableTransition)
        animatable(persistent && !disableTransition)
    }

    useEffect(() => {
        state(({ previous }) => ({
            incrementing: previous < activeIndex,
            previous: disableTransition ? activeIndex : previous,
        }))

        // this is for only updating the incrementing state and should not be fired
        // again if the disableTransition prop is changed
        // eslint-disable-next-line react-hooks/exhaustive-deps
    })

    const onEntered = $(() => {
        state(({ incrementing }) => ({ incrementing, previous: activeIndex }))
    })

    const [ref, refHandler] = useEnsuredRef(forwardedRef)

    useEffect(() => {
        if (!ref() || disableScrollFix) {
            return
        }

        ref().scrollTop = 0
        // don't want it to be triggered if only the disableScrollFix prop has changed
        // since it might be independent from active indexes
        // eslint-disable-next-line react-hooks/exhaustive-deps
    })

    return (
        <div
            {...props}
            ref={refHandler}
            className={[
                block({
                    "slide-left": incrementing,
                    "slide-right": !incrementing,
                }),
                className
            ]}
        >
            {Children.map(children, (child, index) => {
                // if (!isValidElement<TabPanelProps>(child)) {
                //     return child
                // }

                const panel = Children.only(child) as Element
                let labelledBy = panel.getAttribute("aria-labelledby")
                if (!labelledBy && !panel.getAttribute("aria-label") && tabs[index]) {
                    // generally guaranteed to be defined by this point since the TabsManager
                    // will add ids if missing.
                    labelledBy = $$(tabs[index].id)
                }

                return cloneElement(child, {
                    "aria-labelledby": labelledBy,
                    id: `${tabsId}-panel-${index + 1}`,
                    hidden: persistent && index !== activeIndex && index !== previous,
                    temporary: !persistent,
                    transitionIn: index === activeIndex,
                    //@ts-ignore
                    timeout: disableTransition ? 0 : panel[timeout],
                    onEntered: disableTransition ? undefined : onEntered,
                })
            })}
        </div>
    )
}