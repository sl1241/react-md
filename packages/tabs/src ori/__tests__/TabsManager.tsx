import type { ReactElement } from "react";
import { useState } from "react";
import { fireEvent, render } from "@testing-library/react";
import {
  FavoriteSVGIcon,
  HelpSVGIcon,
  TimerSVGIcon,
} from "@react-md/material-icons";
import { Typography } from "@react-md/typography";

import { TabsManager } from "../TabsManager";
import { Tabs } from "../Tabs";
import { TabPanels } from "../TabPanels";
import { TabPanel } from "../TabPanel";
import type { TabConfig } from "../types";

describe("TabsManager", () => {
  it("should work with a real world example", () => {
    const tabs = ["Tab 1", "Tab 2", "Tab 3"];

    function Test(): ReactElement {
      return (
        <TabsManager tabs={tabs} tabsId="basic-usage-tabs">
          <Tabs />
          <TabPanels>
            <TabPanel>
              <Typography type="headline-4">Panel 1</Typography>
            </TabPanel>
            <TabPanel>
              <Typography type="headline-4">Panel 2</Typography>
            </TabPanel>
            <TabPanel>
              <Typography type="headline-4">Panel 3</Typography>
            </TabPanel>
          </TabPanels>
        </TabsManager>
      );
    }
    const { container, getByRole } = render(<Test />);
    expect(container).toMatchSnapshot();
    const tab1 = getByRole("tab", { name: "Tab 1" });
    const tab2 = getByRole("tab", { name: "Tab 2" });
    const tab3 = getByRole("tab", { name: "Tab 3" });

    expect(tab1).toHaveAttribute("aria-selected", "true");
    expect(tab2).toHaveAttribute("aria-selected", "false");
    expect(tab3).toHaveAttribute("aria-selected", "false");

    fireEvent.click(tab2);
    expect(tab1).toHaveAttribute("aria-selected", "false");
    expect(tab2).toHaveAttribute("aria-selected", "true");
    expect(tab3).toHaveAttribute("aria-selected", "false");
  });

  it("should work with object tab configuration and automatic keyboard movement", () => {
    const onActiveIndexChange = jest.fn();
    const tabs: TabConfig[] = [
      { icon: <TimerSVGIcon />, children: "Recent" },
      { icon: <FavoriteSVGIcon />, children: "Favorites" },
    ];

    function Test(): ReactElement {
      return (
        <TabsManager
          stacked
          tabs={tabs}
          tabsId="tabs-id"
          onActiveIndexChange={onActiveIndexChange}
        >
          <Tabs automatic />
          <TabPanels>
            <TabPanel>
              <Typography type="headline-4">Panel 1</Typography>
            </TabPanel>
            <TabPanel>
              <Typography type="headline-4">Panel 2</Typography>
            </TabPanel>
          </TabPanels>
        </TabsManager>
      );
    }

    const { container, getByRole } = render(<Test />);
    const tab1 = getByRole("tab", { name: "Recent" });
    const tab2 = getByRole("tab", { name: "Favorites" });

    expect(onActiveIndexChange).not.toBeCalled();
    expect(tab1).toHaveAttribute("aria-selected", "true");
    expect(tab2).toHaveAttribute("aria-selected", "false");
    expect(container).toMatchSnapshot();

    fireEvent.focus(tab1);
    fireEvent.keyDown(tab1, { key: "ArrowRight" });
    expect(tab1).toHaveAttribute("aria-selected", "false");
    expect(tab2).toHaveAttribute("aria-selected", "true");
    expect(onActiveIndexChange).toBeCalledWith(1);
  });

  it("should work with disabled tabs", () => {
    const tabs: TabConfig[] = [
      { icon: <TimerSVGIcon />, children: "Recent" },
      { icon: <FavoriteSVGIcon />, children: "Favorites" },
      { icon: <HelpSVGIcon />, disabled: true, children: "Other" },
    ];

    function Test(): ReactElement {
      return (
        <TabsManager stacked tabs={tabs} tabsId="tabs-id">
          <Tabs automatic />
          <TabPanels>
            <TabPanel>
              <Typography type="headline-4">Panel 1</Typography>
            </TabPanel>
            <TabPanel>
              <Typography type="headline-4">Panel 2</Typography>
            </TabPanel>
            <TabPanel>
              <Typography type="headline-4">Panel 3</Typography>
            </TabPanel>
          </TabPanels>
        </TabsManager>
      );
    }

    const { container, getByRole } = render(<Test />);
    const tab1 = getByRole("tab", { name: "Recent" });
    const tab2 = getByRole("tab", { name: "Favorites" });
    const tab3 = getByRole("tab", { name: "Other" });

    expect(tab1).toHaveAttribute("aria-selected", "true");
    expect(tab2).toHaveAttribute("aria-selected", "false");
    expect(tab3).toHaveAttribute("aria-selected", "false");
    expect(container).toMatchSnapshot();

    fireEvent.click(tab3);
    expect(tab1).toHaveAttribute("aria-selected", "true");
    expect(tab2).toHaveAttribute("aria-selected", "false");
    expect(tab3).toHaveAttribute("aria-selected", "false");
  });

  it("should be controllable", () => {
    const tabs = ["Tab 1", "Tab 2"] as const;
    function Test(): ReactElement {
      const [activeIndex, setActiveIndex] = useState(1);

      return (
        <TabsManager
          tabs={tabs}
          tabsId="tabs-id"
          activeIndex={activeIndex}
          onActiveIndexChange={setActiveIndex}
        >
          <Tabs />
          <TabPanels>
            <TabPanel>
              <Typography type="headline-4">Panel 1</Typography>
            </TabPanel>
            <TabPanel>
              <Typography type="headline-4">Panel 2</Typography>
            </TabPanel>
          </TabPanels>
        </TabsManager>
      );
    }

    const { getByRole } = render(<Test />);
    const tab1 = getByRole("tab", { name: "Tab 1" });
    const tab2 = getByRole("tab", { name: "Tab 2" });
    expect(tab1).toHaveAttribute("aria-selected", "false");
    expect(tab2).toHaveAttribute("aria-selected", "true");

    fireEvent.click(tab1);
    expect(tab1).toHaveAttribute("aria-selected", "true");
    expect(tab2).toHaveAttribute("aria-selected", "false");
  });
});
