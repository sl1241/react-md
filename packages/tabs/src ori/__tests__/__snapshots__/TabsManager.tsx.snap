// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsManager should work with a real world example 1`] = `
<div>
  <div
    aria-orientation="horizontal"
    class="rmd-tabs rmd-tabs--left rmd-tabs--animate"
    id="basic-usage-tabs"
    role="tablist"
    style="--rmd-tab-width: 33.333333333333336%; --rmd-tab-offset: calc(0 * 33.333333333333336%);"
  >
    <button
      aria-controls="basic-usage-tabs-panel-1"
      aria-selected="true"
      class="rmd-tab rmd-tab--active"
      id="basic-usage-tabs-tab-1"
      role="tab"
      type="button"
    >
      <span
        class="rmd-tab__content"
      >
        Tab 1
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
    <button
      aria-selected="false"
      class="rmd-tab"
      id="basic-usage-tabs-tab-2"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <span
        class="rmd-tab__content"
      >
        Tab 2
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
    <button
      aria-selected="false"
      class="rmd-tab"
      id="basic-usage-tabs-tab-3"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <span
        class="rmd-tab__content"
      >
        Tab 3
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
  </div>
  <div
    class="rmd-tab-panels rmd-tab-panels--slide-right"
  >
    <div
      aria-labelledby="basic-usage-tabs-tab-1"
      class="rmd-tab-panel"
      id="basic-usage-tabs-panel-1"
      role="tabpanel"
    >
      <h4
        class="rmd-typography rmd-typography--headline-4"
      >
        Panel 1
      </h4>
    </div>
  </div>
</div>
`;

exports[`TabsManager should work with disabled tabs 1`] = `
<div>
  <div
    aria-orientation="horizontal"
    class="rmd-tabs rmd-tabs--left rmd-tabs--animate"
    id="tabs-id"
    role="tablist"
    style="--rmd-tab-width: 33.333333333333336%; --rmd-tab-offset: calc(0 * 33.333333333333336%);"
  >
    <button
      aria-controls="tabs-id-panel-1"
      aria-selected="true"
      class="rmd-tab rmd-tab--active rmd-tab--stacked"
      id="tabs-id-tab-1"
      role="tab"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="rmd-icon rmd-icon--svg rmd-icon--above"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 1H9v2h6V1zm-4 13h2V8h-2v6zm8.03-6.61l1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42C16.07 4.74 14.12 4 12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9 9-4.03 9-9c0-2.12-.74-4.07-1.97-5.61zM12 20c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
        />
      </svg>
      <span
        class="rmd-tab__content"
      >
        Recent
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
    <button
      aria-selected="false"
      class="rmd-tab rmd-tab--stacked"
      id="tabs-id-tab-2"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="rmd-icon rmd-icon--svg rmd-icon--above"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
        />
      </svg>
      <span
        class="rmd-tab__content"
      >
        Favorites
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
    <button
      aria-selected="false"
      class="rmd-tab rmd-tab--stacked"
      disabled=""
      id="tabs-id-tab-3"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="rmd-icon rmd-icon--svg rmd-icon--above"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
        />
      </svg>
      <span
        class="rmd-tab__content"
      >
        Other
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
  </div>
  <div
    class="rmd-tab-panels rmd-tab-panels--slide-right"
  >
    <div
      aria-labelledby="tabs-id-tab-1"
      class="rmd-tab-panel"
      id="tabs-id-panel-1"
      role="tabpanel"
    >
      <h4
        class="rmd-typography rmd-typography--headline-4"
      >
        Panel 1
      </h4>
    </div>
  </div>
</div>
`;

exports[`TabsManager should work with object tab configuration and automatic keyboard movement 1`] = `
<div>
  <div
    aria-orientation="horizontal"
    class="rmd-tabs rmd-tabs--left rmd-tabs--animate"
    id="tabs-id"
    role="tablist"
    style="--rmd-tab-width: 50%; --rmd-tab-offset: calc(0 * 50%);"
  >
    <button
      aria-controls="tabs-id-panel-1"
      aria-selected="true"
      class="rmd-tab rmd-tab--active rmd-tab--stacked"
      id="tabs-id-tab-1"
      role="tab"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="rmd-icon rmd-icon--svg rmd-icon--above"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 1H9v2h6V1zm-4 13h2V8h-2v6zm8.03-6.61l1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42C16.07 4.74 14.12 4 12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9 9-4.03 9-9c0-2.12-.74-4.07-1.97-5.61zM12 20c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
        />
      </svg>
      <span
        class="rmd-tab__content"
      >
        Recent
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
    <button
      aria-selected="false"
      class="rmd-tab rmd-tab--stacked"
      id="tabs-id-tab-2"
      role="tab"
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="rmd-icon rmd-icon--svg rmd-icon--above"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
        />
      </svg>
      <span
        class="rmd-tab__content"
      >
        Favorites
      </span>
      <span
        class="rmd-ripple-container"
      />
    </button>
  </div>
  <div
    class="rmd-tab-panels rmd-tab-panels--slide-right"
  >
    <div
      aria-labelledby="tabs-id-tab-1"
      class="rmd-tab-panel"
      id="tabs-id-panel-1"
      role="tabpanel"
    >
      <h4
        class="rmd-typography rmd-typography--headline-4"
      >
        Panel 1
      </h4>
    </div>
  </div>
</div>
`;
